node/src/transaction_pool/candidate/transaction_pool_candidate_reducer.rs

```diff
@@ -90,38 +89,53 @@ impl TransactionPoolCandidatesState {
             } => {
                 state.transaction_received(meta.time(), *peer_id, transaction.clone());
             }
+            TransactionPoolCandidateAction::Libp2pTransactionsReceived {
+                peer_id,
+                transactions,
+                message_id,
+            } => {
+                state.transactions_received(
+                    meta.time(),
+                    *peer_id,
+                    transactions.clone(),
+                    *message_id,
+                );
+            }
             TransactionPoolCandidateAction::VerifyNext => {
                 let (dispatcher, global_state) = state_context.into_dispatcher_and_state();
 
                 let batch = global_state
                     .transaction_pool
                     .candidates
                     .get_batch_to_verify();
-                let Some((peer_id, batch)) = batch else {
+                let Some((peer_id, batch, from_source)) = batch else {
                     return;
                 };
 
                 let transaction_hashes = batch.iter().map(|tx| tx.hash().clone()).collect();
                 dispatcher.push(TransactionPoolAction::StartVerify {
                     commands: batch.into_iter().collect(),
-                    from_source: TransactionPoolMessageSource::None,
+                    from_source,
                 });
                 dispatcher.push(TransactionPoolCandidateAction::VerifyPending {
                     peer_id,
                     transaction_hashes,
                     verify_id: (),
+                    from_source,
                 });
             }
             TransactionPoolCandidateAction::VerifyPending {
                 peer_id,
                 transaction_hashes,
                 verify_id,
+                from_source,
             } => {
                 state.verify_pending(meta.time(), peer_id, *verify_id, transaction_hashes);
                 let dispatcher = state_context.into_dispatcher();
                 dispatcher.push(TransactionPoolCandidateAction::VerifySuccess {
```

Comment:
Something else I am confused about is that we set the verification state to pending then immediately dispatch a success.

Reply:
Overall the flow looks right, because these stop being candidates, and we want to cleanup that state. It is just that the naming and the way it is modeled here is confusing. Something to improve later.